c_binding/lsm_local_disk.c

```diff
@@ -804,3 +795,75 @@ static int _sysfs_sas_addr_get(const char *blk_name, char *tp_sas_addr)
 
     return rc;
 }
+
+static int _sas_addr_get(char *err_msg, const char *disk_path,
+                         char *tp_sas_addr)
+{
+    int rc = LSM_ERR_OK;
+    int fd = -1;
+
+    assert(disk_path != NULL);
+    assert(tp_sas_addr != NULL);
+
+    memset(tp_sas_addr, 0, _SG_T10_SPL_SAS_ADDR_LEN);
+
+    /* TODO(Gris Ge): Add support of NVMe enclosure */
+
+    /* Try use sysfs first to get SAS address. */
+    if ((strlen(disk_path) > strlen("/dev/")) &&
+        (strncmp(disk_path, "/dev/", strlen("/dev/")) == 0) &&
+        (strncmp(disk_path + strlen("/dev/"), "sd", strlen("sd")) == 0))
+        _sysfs_sas_addr_get(disk_path + strlen("/dev/"), tp_sas_addr);
+
+    if (tp_sas_addr[0] == '\0') {
+        _good(_sg_io_open_ro(err_msg, disk_path, &fd), rc, out);
+        _good(_sg_tp_sas_addr_of_disk(err_msg, fd, tp_sas_addr), rc, out);
+    }
+
+ out:
+    if (fd >= 0)
+        close(fd);
+    return rc;
+
+}
+
+int LSM_DLL_EXPORT lsm_local_disk_led_status_get(const char *disk_path,
+                                                 uint32_t *led_status,
+                                                 lsm_error **lsm_err)
+{
+    int rc = LSM_ERR_OK;
+    char err_msg[_LSM_ERR_MSG_LEN];
+    char tp_sas_addr[_SG_T10_SPL_SAS_ADDR_LEN];
+    struct _ses_dev_slot_status status;
+
+    _lsm_err_msg_clear(err_msg);
+
+    _good(_check_null_ptr(err_msg, 3 /* arg_count */, disk_path, led_status,
+                          lsm_err),
+          rc, out);
+
+    _good(_sas_addr_get(err_msg, disk_path, tp_sas_addr), rc, out);
+
+    _good(_ses_status_get(err_msg, tp_sas_addr, &status), rc, out);
+
+    *led_status = 0;
+
+    if (status.fault_reqstd || status.fault_sensed)
+        *led_status |= LSM_DISK_LED_STATUS_FAULT_ON;
+    else
+        *led_status |= LSM_DISK_LED_STATUS_FAULT_OFF;
+
+    if (status.ident)
+        *led_status |= LSM_DISK_LED_STATUS_IDENT_ON;
+    else
+        *led_status |= LSM_DISK_LED_STATUS_IDENT_OFF;
+
+ out:
+    if (rc != LSM_ERR_OK) {
+        if (led_status != NULL)
```

Comment:
I think it might be cleaner if you add an additional `goto` labeled `badargs` or something like it below `out` label and then change the first good to jump to it instead.  That way you don't need to do the if checks for `led_status` and `lsm_err`.  This is subjective and I could see people leaning each way.